Shurikens

emanuelmch opened the pull request:
The code needs some cleanup, and the commits... ok, don't get me started on the commits. BUT IT MOTHERFREAKING WORKS, AND THAT'S ALL THAT MATTERS!! I'm happy :)
Update: Now it's good. Now I think I can merge this in. The only thing I'm still not satisfied about is my mechanism for handling multiple possible solutions to the same shuriken (Introduced in https://github.com/emanuelmch/puzzles/pull/10/commits/630d553088579dac480c1775c4fab474936b9fc7)

emanuelmch committed:
Add base files for Shurikens

No solver implementation yet.

emanuelmch committed:
Add a very basic solver

emanuelmch committed:
Add a couple of real Shurikens

I've also found more than one valid solution, so I also had to create a
mechanism for accepting various different solutions as valid answers.
Not only that, but it also turns out this can take a while to calculate,
so I had to make sure it was behind a parameter wall.

emanuelmch committed:
Add the TwoBitStorage class

This is used in Breadth Search Solver for a massive memory optimization

emanuelmch committed:
Start separating into multiple ShurikenData

emanuelmch committed:
Add an easy shuriken

herculesjr commented on src/common/twobitstorage.cpp:
```diff
@@ -0,0 +1,56 @@
+u_int8_t TwoBitStorage::operator[](size_t i) const {
+  assert(i < _size);
+  auto shift = (MAX_SIZE - 1 - i) * 2;
+  return (internal & (static_cast<u_int64_t >(3) << shift)) >> shift;
```
```suggestion
  return (internal & (static_cast<u_int64_t>(3) << shift)) >> shift;
```

emanuelmch committed:
Optimize solver by pre-allocating memory

emanuelmch committed:
Treat both sides of the shuriken as interchangeable

This leads to a massive performance improvement, as we now only have to
care about half of the shurikens/nodes we needed to. The run times have
not decrease in half (because we weren't hitting all the nodes we now
ignore), but it's still very significant.
A side effect was that the model unit grew too big, so I moved most of the code
out of the header file into a new compilation unit.

emanuelmch committed:
Add a second Shuriken solver

emanuelmch commented on src/main.cpp:
```diff
@@ -93,25 +99,32 @@ bool solveCPic() {
-bool solveShurikens() {
-  Shurikens::BreadthSearchSolver bruteSolver;
+bool solveShurikens(bool fullRun) {
+  Shurikens::BreadthSearchSolver solver;
```
this should've been called "solver" from the start
